python/shark_turbine/kernel/_support/indexing.py

```diff
@@ -569,3 +555,134 @@ def __enter__(self) -> "IndexingContext":
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         context.pop(IndexingContext, self)
+
+
+###############################################################################
+# Symbolic index value type.
+# This type is used in place of `int` to track a symbolic value during
+# traced execution. The type does not carry
+###############################################################################
+
+
+class IndexRelation(ABC):
+    """ABC for assumptions that can be made about an index value."""
+
+    __slots__ = []
+
+
+class EqualRelation(IndexRelation):
+    """An index assumption that can take a single symbolic value."""
+
+    __slots__ = ["eq_expr"]
+
+    def __init__(self, eq_expr: IndexExpr):
+        self.eq_expr = eq_expr
+
+    def __eq__(self, other):
+        if not isinstance(other, EqualRelation):
+            return False
+        return self.eq_expr == other.eq_expr
+
+    def __repr__(self):
+        expr = self.eq_expr
+        if isinstance(expr, IndexSymbol):
+            return f"=={expr}"
+        else:
+            return f"==({expr})"
+
+
+class BoundedRelation(IndexRelation):
```

Comment:
I'm not sure if the Index value itself should be tracking this. This seems like something that would come up as an FX analysis pass instead.

Reply:
Yeah, I don't love where this part got to...

Reply:
Added a TODO to factor this out.